File: backend/app/infrastructure/persistence/sqlite/repositories/local_item_repository.py
"""
SQLite Local Item Repository — مشروع «مُعين» (Mouin)
Implements IItemRepository with Workspace Scoping & Cascade Deletes.
"""

import sqlite3
from typing import List, Optional
from backend.app.application.ports.repositories import IItemRepository
from backend.app.domain.entities.item import Item
from backend.app.domain.value_objects.identity import EntityId, WorkspaceId
from backend.app.domain.value_objects.types import ItemType
from backend.app.infrastructure.persistence.sqlite.mappers.local_item_mapper import SqliteItemMapper

def iso(dt) -> Optional[str]:
    return dt.isoformat() if dt else None

class SqliteItemRepository(IItemRepository):
    def __init__(self, connection: sqlite3.Connection):
        self.connection = connection

    def save(self, item: Item) -> None:
        # 1. Upsert local_items root
        sql_item = """
            INSERT INTO local_items (
                id, workspace_id, item_type, title, summary, category_id,
                privacy_classification, temporal_original_expression, temporal_resolved_at,
                temporal_timezone, temporal_locale, temporal_calendar,
                created_by_installation_id, created_at, updated_at, deleted_at, entity_version
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            ON CONFLICT (id) DO UPDATE SET
                title = excluded.title,
                summary = excluded.summary,
                category_id = excluded.category_id,
                privacy_classification = excluded.privacy_classification,
                temporal_original_expression = excluded.temporal_original_expression,
                temporal_resolved_at = excluded.temporal_resolved_at,
                updated_at = excluded.updated_at,
                deleted_at = excluded.deleted_at,
                entity_version = excluded.entity_version;
        """
        self.connection.execute(sql_item, (
            str(item.id), str(item.workspace_id), item.item_type.value, item.title, item.summary,
            str(item.category_id) if item.category_id else None,
            item.privacy_classification.value, item.temporal_original_expression,
            iso(item.temporal_resolved_at), item.temporal_timezone, item.temporal_locale, item.temporal_calendar,
            str(item.created_by_installation_id) if item.created_by_installation_id else None,
            iso(item.created_at), iso(item.updated_at), iso(item.deleted_at), item.entity_version
        ))

        # 2. Upsert specialized subtype table
        if item.item_type == ItemType.TASK and item.task_detail:
            sql_task = """
                INSERT INTO local_tasks (item_id, due_date, priority, status, completed_at, estimated_duration_minutes)
                VALUES (?, ?, ?, ?, ?, ?)
                ON CONFLICT (item_id) DO UPDATE SET
                    due_date = excluded.due_date,
                    priority = excluded.priority,
                    status = excluded.status,
                    completed_at = excluded.completed_at,
                    estimated_duration_minutes = excluded.estimated_duration_minutes;
            """
            self.connection.execute(sql_task, (
                str(item.id), iso(item.task_detail.due_date), item.task_detail.priority.value,
                item.task_detail.status.value, iso(item.task_detail.completed_at), item.task_detail.estimated_duration_minutes
            ))
        elif item.item_type == ItemType.APPOINTMENT and item.appointment_detail:
            sql_appt = """
                INSERT INTO local_appointments (item_id, start_time, end_time, location, all_day, timezone)
                VALUES (?, ?, ?, ?, ?, ?)
                ON CONFLICT (item_id) DO UPDATE SET
                    start_time = excluded.start_time,
                    end_time = excluded.end_time,
                    location = excluded.location,
                    all_day = excluded.all_day,
                    timezone = excluded.timezone;
            """
            self.connection.execute(sql_appt, (
                str(item.id), iso(item.appointment_detail.start_time), iso(item.appointment_detail.end_time),
                item.appointment_detail.location, 1 if item.appointment_detail.all_day else 0, item.appointment_detail.timezone
            ))
        elif item.item_type == ItemType.NOTE and item.note_detail:
            sql_note = """
                INSERT INTO local_notes (item_id, content, content_format)
                VALUES (?, ?, ?)
                ON CONFLICT (item_id) DO UPDATE SET
                    content = excluded.content,
                    content_format = excluded.content_format;
            """
            self.connection.execute(sql_note, (
                str(item.id), item.note_detail.content, item.note_detail.content_format
            ))
        elif item.item_type == ItemType.DOCUMENT and item.document_detail:
            sql_doc = """
                INSERT INTO local_documents (item_id, document_type, issue_date, expiry_date, document_number, issuing_authority)
                VALUES (?, ?, ?, ?, ?, ?)
                ON CONFLICT (item_id) DO UPDATE SET
                    document_type = excluded.document_type,
                    issue_date = excluded.issue_date,
                    expiry_date = excluded.expiry_date,
                    document_number = excluded.document_number,
                    issuing_authority = excluded.issuing_authority;
            """
            self.connection.execute(sql_doc, (
                str(item.id), item.document_detail.document_type,
                iso(item.document_detail.issue_date), iso(item.document_detail.expiry_date),
                item.document_detail.document_number, item.document_detail.issuing_authority
            ))

    def get_by_id(self, workspace_id: WorkspaceId, item_id: EntityId) -> Optional[Item]:
        cursor = self.connection.cursor()
        cursor.execute("SELECT * FROM local_items WHERE workspace_id = ? AND id = ?;", (str(workspace_id), str(item_id)))
        item_row = cursor.fetchone()
        if not item_row:
            return None

        item_type = item_row['item_type']
        table_map = {
            "task": "local_tasks", "appointment": "local_appointments", "note": "local_notes",
            "document": "local_documents", "debt": "local_debts", "shopping": "local_shopping_lists"
        }
        subtype_row = None
        if item_type in table_map:
            tname = table_map[item_type]
            cursor.execute(f"SELECT * FROM {tname} WHERE item_id = ?;", (str(item_id),))
            subtype_row = cursor.fetchone()

        return SqliteItemMapper.to_domain(item_row, subtype_row)

    def list_by_workspace(self, workspace_id: WorkspaceId, limit: int = 50, offset: int = 0) -> List[Item]:
        cursor = self.connection.cursor()
        cursor.execute(
            "SELECT id FROM local_items WHERE workspace_id = ? AND deleted_at IS NULL ORDER BY created_at DESC LIMIT ? OFFSET ?;",
            (str(workspace_id), limit, offset)
        )
        rows = cursor.fetchall()
        return [self.get_by_id(workspace_id, EntityId(str(r['id']))) for r in rows if r]

File: mobile/lib/infrastructure/repositories/local_item_repository.dart
// Local SQLite Repositories — مشروع «مُعين» (Mouin)
import 'package:mouin/core/result/result.dart';
import 'package:mouin/core/errors/failures.dart';
import 'package:mouin/core/utils/arabic_normalizer.dart';
import 'package:mouin/domain/entities/item.dart';
import 'package:mouin/domain/entities/debt.dart';
import 'package:mouin/domain/entities/reminder.dart';
import 'package:mouin/domain/repositories/i_item_repository.dart';
import 'package:mouin/domain/value_objects/types.dart';
import 'package:mouin/domain/value_objects/money.dart';
import 'package:mouin/infrastructure/database/local_sqlite_db.dart';

class LocalItemRepository implements IItemRepository {
  final LocalSqliteDb db;
  LocalItemRepository(this.db);

  @override
  Future<Result<void, Failure>> save(Item item) async {
    // 1. Upsert root item
    db.items[item.id] = {
      'id': item.id,
      'workspace_id': item.workspaceId,
      'item_type': item.itemType.name,
      'title': item.title,
      'summary': item.summary,
      'privacy_classification': item.privacy.name,
      'category_id': item.categoryId,
      'created_at': item.createdAt.toIso8601String(),
      'updated_at': item.updatedAt.toIso8601String(),
      'deleted_at': item.deletedAt?.toIso8601String(),
      'entity_version': item.entityVersion,
    };

    // 2. Upsert subtype details
    if (item.taskDetail != null) {
      db.tasks[item.id] = {
        'item_id': item.id,
        'due_date': item.taskDetail!.dueDate?.toIso8601String(),
        'priority': item.taskDetail!.priority.name,
        'status': item.taskDetail!.status.name,
        'completed_at': item.taskDetail!.completedAt?.toIso8601String(),
        'estimated_duration_minutes': item.taskDetail!.estimatedDurationMinutes,
      };
    }
    if (item.noteDetail != null) {
      db.notes[item.id] = {
        'item_id': item.id,
        'content': item.noteDetail!.content,
        'content_format': item.noteDetail!.contentFormat,
        'word_count': item.noteDetail!.wordCount,
      };
    }
    if (item.appointmentDetail != null) {
      db.appointments[item.id] = {
        'item_id': item.id,
        'start_time': item.appointmentDetail!.startTime.toIso8601String(),
        'end_time': item.appointmentDetail!.endTime?.toIso8601String(),
        'location': item.appointmentDetail!.location,
        'calendar_event_id': item.appointmentDetail!.calendarEventId,
        'all_day': item.appointmentDetail!.allDay ? 1 : 0,
        'timezone': item.appointmentDetail!.timezone,
      };
    }
    if (item.documentDetail != null) {
      db.documents[item.id] = {
        'item_id': item.id,
        'document_type': item.documentDetail!.documentType,
        'document_number': item.documentDetail!.documentNumber,
        'issuing_authority': item.documentDetail!.issuingAuthority,
        'issue_date': item.documentDetail!.issueDate?.toIso8601String(),
        'expiry_date': item.documentDetail!.expiryDate?.toIso8601String(),
        'storage_path': item.documentDetail!.storagePath,
        'mime_type': item.documentDetail!.mimeType,
        'file_size_bytes': item.documentDetail!.fileSizeBytes,
      };
    }
    return const Result.success(null);
  }

  @override
  Future<Result<Item?, Failure>> getById(String workspaceId, String id) async {
    final row = db.items[id];
    if (row == null || row['workspace_id'] != workspaceId) {
      return const Result.success(null);
    }

    TaskDetail? taskDetail;
    final tRow = db.tasks[id];
    if (tRow != null) {
      taskDetail = TaskDetail(
        dueDate: tRow['due_date'] != null ? DateTime.parse(tRow['due_date']) : null,
        priority: Priority.values.byName(tRow['priority']),
        status: TaskStatus.values.byName(tRow['status']),
        completedAt: tRow['completed_at'] != null ? DateTime.parse(tRow['completed_at']) : null,
        estimatedDurationMinutes: tRow['estimated_duration_minutes'],
      );
    }

    NoteDetail? noteDetail;
    final nRow = db.notes[id];
    if (nRow != null) {
      noteDetail = NoteDetail(
        content: nRow['content'] ?? '',
        contentFormat: nRow['content_format'] ?? 'plain_text',
        wordCount: nRow['word_count'],
      );
    }

    AppointmentDetail? apptDetail;
    final aRow = db.appointments[id];
    if (aRow != null) {
      apptDetail = AppointmentDetail(
        startTime: DateTime.parse(aRow['start_time']),
        endTime: aRow['end_time'] != null ? DateTime.parse(aRow['end_time']) : null,
        location: aRow['location'],
        calendarEventId: aRow['calendar_event_id'],
        allDay: aRow['all_day'] == 1,
        timezone: aRow['timezone'] ?? 'Asia/Aden',
      );
    }

    DocumentDetail? docDetail;
    final dRow = db.documents[id];
    if (dRow != null) {
      docDetail = DocumentDetail(
        documentType: dRow['document_type'] ?? 'general',
        documentNumber: dRow['document_number'],
        issuingAuthority: dRow['issuing_authority'],
        issueDate: dRow['issue_date'] != null ? DateTime.parse(dRow['issue_date']) : null,
        expiryDate: dRow['expiry_date'] != null ? DateTime.parse(dRow['expiry_date']) : null,
        storagePath: dRow['storage_path'],
        mimeType: dRow['mime_type'],
        fileSizeBytes: dRow['file_size_bytes'],
      );
    }

    final item = Item(
      id: row['id'],
      workspaceId: row['workspace_id'],
      itemType: ItemType.values.byName(row['item_type']),
      title: row['title'],
      summary: row['summary'],
      privacy: PrivacyClassification.values.byName(row['privacy_classification']),
      categoryId: row['category_id'],
      taskDetail: taskDetail,
      noteDetail: noteDetail,
      appointmentDetail: apptDetail,
      documentDetail: docDetail,
      createdAt: DateTime.parse(row['created_at']),
      updatedAt: DateTime.parse(row['updated_at']),
      deletedAt: row['deleted_at'] != null ? DateTime.parse(row['deleted_at']) : null,
      entityVersion: row['entity_version'] ?? 1,
    );
    return Result.success(item);
  }

  @override
  Future<Result<List<Item>, Failure>> listByWorkspace(String workspaceId, {int limit = 50, int offset = 0}) async {
    final list = <Item>[];
    for (final row in db.items.values) {
      if (row['workspace_id'] == workspaceId && row['deleted_at'] == null) {
        final res = await getById(workspaceId, row['id']);
        if (res.isSuccess && res.value != null) {
          list.add(res.value!);
        }
      }
    }
    return Result.success(list);
  }

  @override
  Future<Result<List<Item>, Failure>> searchArabic(String workspaceId, String query) async {
    final normalizedQuery = ArabicNormalizer.normalize(query);
    final results = <Item>[];
    for (final row in db.items.values) {
      if (row['workspace_id'] == workspaceId && row['deleted_at'] == null) {
        final normTitle = ArabicNormalizer.normalize(row['title'] ?? '');
        final normSummary = ArabicNormalizer.normalize(row['summary'] ?? '');
        if (normTitle.contains(normalizedQuery) || normSummary.contains(normalizedQuery)) {
          final itemRes = await getById(workspaceId, row['id']);
          if (itemRes.isSuccess && itemRes.value != null) {
            results.add(itemRes.value!);
          }
        }
      }
    }
    return Result.success(results);
  }
}

class LocalDebtRepository implements IDebtRepository {
  final LocalSqliteDb db;
  LocalDebtRepository(this.db);

  @override
  Future<Result<void, Failure>> save(Debt debt) async {
    db.debts[debt.id] = {
      'id': debt.id,
      'workspace_id': debt.workspaceId,
      'person_id': debt.personId,
      'debt_type': debt.debtType.name,
      'total_amount': debt.totalAmount.toDecimalString(),
      'currency': debt.totalAmount.currency,
      'status': debt.status.name,
      'due_date': debt.dueDate?.toIso8601String(),
      'created_at': debt.createdAt.toIso8601String(),
      'updated_at': debt.updatedAt.toIso8601String(),
      'deleted_at': debt.deletedAt?.toIso8601String(),
      'entity_version': debt.entityVersion,
    };

    for (final tx in debt.transactions) {
      db.debtTransactions[tx.id] = {
        'id': tx.id,
        'debt_id': tx.debtId,
        'workspace_id': tx.workspaceId,
        'transaction_type': tx.transactionType.name,
        'amount': tx.amount.toDecimalString(),
        'currency': tx.amount.currency,
        'transaction_date': tx.transactionDate.toIso8601String(),
        'notes': tx.notes,
        'reference_transaction_id': tx.referenceTransactionId,
        'created_at': tx.createdAt.toIso8601String(),
      };
    }
    return const Result.success(null);
  }

  @override
  Future<Result<Debt?, Failure>> getById(String workspaceId, String id) async {
    final row = db.debts[id];
    if (row == null || row['workspace_id'] != workspaceId) {
      return const Result.success(null);
    }

    final txList = <DebtTransaction>[];
    for (final txRow in db.debtTransactions.values) {
      if (txRow['debt_id'] == id && txRow['workspace_id'] == workspaceId) {
        txList.add(DebtTransaction(
          id: txRow['id'],
          debtId: txRow['debt_id'],
          workspaceId: txRow['workspace_id'],
          transactionType: DebtTransactionType.values.byName(txRow['transaction_type']),
          amount: Money.fromDecimalString(txRow['amount'], currency: txRow['currency'] ?? 'YER'),
          transactionDate: DateTime.parse(txRow['transaction_date']),
          notes: txRow['notes'],
          referenceTransactionId: txRow['reference_transaction_id'],
          createdAt: DateTime.parse(txRow['created_at']),
        ));
      }
    }

    final debt = Debt(
      id: row['id'],
      workspaceId: row['workspace_id'],
      personId: row['person_id'],
      debtType: DebtType.values.byName(row['debt_type']),
      totalAmount: Money.fromDecimalString(row['total_amount'], currency: row['currency'] ?? 'YER'),
      status: DebtStatus.values.byName(row['status']),
      dueDate: row['due_date'] != null ? DateTime.parse(row['due_date']) : null,
      transactions: txList,
      createdAt: DateTime.parse(row['created_at']),
      updatedAt: DateTime.parse(row['updated_at']),
      deletedAt: row['deleted_at'] != null ? DateTime.parse(row['deleted_at']) : null,
      entityVersion: row['entity_version'],
    );
    return Result.success(debt);
  }

  @override
  Future<Result<List<Debt>, Failure>> listByWorkspace(String workspaceId) async {
    final list = <Debt>[];
    for (final row in db.debts.values) {
      if (row['workspace_id'] == workspaceId && row['deleted_at'] == null) {
        final res = await getById(workspaceId, row['id']);
        if (res.isSuccess && res.value != null) {
          list.add(res.value!);
        }
      }
    }
    return Result.success(list);
  }
}

class LocalReminderRepository implements IReminderRepository {
  final LocalSqliteDb db;
  LocalReminderRepository(this.db);

  @override
  Future<Result<void, Failure>> saveRule(ReminderRule rule) async {
    db.reminderRules[rule.id] = {
      'id': rule.id,
      'workspace_id': rule.workspaceId,
      'item_id': rule.itemId,
      'trigger_type': rule.triggerType.name,
      'trigger_time': rule.triggerTime?.toIso8601String(),
      'offset_minutes': rule.offsetMinutes,
      'rrule': rule.rrule,
      'is_active': rule.isActive ? 1 : 0,
    };
    return const Result.success(null);
  }

  @override
  Future<Result<ReminderRule?, Failure>> getRuleById(String workspaceId, String id) async {
    final row = db.reminderRules[id];
    if (row == null || row['workspace_id'] != workspaceId) {
      return const Result.success(null);
    }
    final rule = ReminderRule(
      id: row['id'],
      workspaceId: row['workspace_id'],
      itemId: row['item_id'],
      triggerType: ReminderTriggerType.values.byName(row['trigger_type']),
      triggerTime: row['trigger_time'] != null ? DateTime.parse(row['trigger_time']) : null,
      offsetMinutes: row['offset_minutes'],
      rrule: row['rrule'],
      isActive: row['is_active'] == 1,
    );
    return Result.success(rule);
  }

  @override
  Future<Result<void, Failure>> saveInstance(ReminderInstance instance) async {
    db.reminderInstances[instance.id] = {
      'id': instance.id,
      'rule_id': instance.ruleId,
      'item_id': instance.itemId,
      'workspace_id': instance.workspaceId,
      'occurrence_key': instance.occurrenceKey,
      'scheduled_time': instance.scheduledTime.toIso8601String(),
      'status': instance.status.name,
      'snoozed_until': instance.snoozedUntil?.toIso8601String(),
    };
    return const Result.success(null);
  }
}

class LocalOutboxRepository implements IOutboxRepository {
  final LocalSqliteDb db;
  LocalOutboxRepository(this.db);

  @override
  Future<Result<void, Failure>> enqueue({
    required String operationId,
    required String entityType,
    required String entityId,
    required String operation,
    required Map<String, dynamic> payload,
    int baseVersion = 1,
  }) async {
    db.outbox[operationId] = {
      'operation_id': operationId,
      'entity_type': entityType,
      'entity_id': entityId,
      'operation': operation,
      'payload': payload,
      'base_version': baseVersion,
      'status': 'pending',
      'attempt_count': 0,
      'created_at': DateTime.now().toUtc().toIso8601String(),
    };
    return const Result.success(null);
  }

  @override
  Future<Result<List<Map<String, dynamic>>, Failure>> getPendingOperations({int limit = 50}) async {
    final list = db.outbox.values.where((op) => op['status'] == 'pending').toList();
    return Result.success(list.take(limit).toList());
  }

  @override
  Future<Result<void, Failure>> markCompleted(String operationId) async {
    db.outbox.remove(operationId);
    return const Result.success(null);
  }
}
